### ISSUE TITLE
why does SUSE need its own style guide?

### ISSUE THREAD

**Issue by Ana06** [SUSE/style-guides#18]
Why does SUSE need a custom style guide? Why not following the community conventions (Rubocop defaults) which more likely are used in most projects out there? 

Every Rubocop version has different defaults as they change over time... how does this work with the single Rubocop file that there is in this repo?

How is it decided what are SUSE style guide? When Rubocop add new rules, who decide if this should be part of SUSE style guide?

As Rubocop is community maintained and the default configuration can be changed depending on what Ruby community think, did someone raised the concerns of this guide there?

**Comment by mssola** [comment]
Hi, maintainer here. This is a question I've been wondering myself, and it has sparked a couple of discussions already in the containers team.

> Every Rubocop version has different defaults as they change over time... how does this work with the single Rubocop file that there is in this repo?

It really doesn't. From time to time we have to update this to catch up with upstream (see #17), which is troublesome.

> How is it decided what are SUSE style guide? When Rubocop add new rules, who decide if this should be part of SUSE style guide?

Theoretically anyone from SUSE that cares about this. That being said, in practice not much time has been spent on this style guide, since everyone has more pressing concerns (see PRs that have stalled for a long time now...). So, it's a problem...

> As Rubocop is community maintained and the default configuration can be changed depending on what Ruby community think, did someone raised the concerns of this guide there?

That's a good question indeed, and I think it's really the way to go. The discussions we've been having in the containers team is that we should stick as close as possible to upstream. That being said, I still see a value of having a small set of rules that are common across teams.

So, personally, I'd perform the following changes to this style guide:

- Keep stuff that we agree and the community guide has no strong opinions. For example: [string literals](https://github.com/bbatsov/ruby-style-guide#consistent-string-literals). The community guide says that both options are valid, and when we discussed about it at SUSE, the preferred options seemed double quotes (which is something I didn't agree much at first, but now I got used to it 😅).
- Some options have to be re-adjusted. For example, as you say, the community has changed some opinions on the guide, and so stuff like `AllowURI` and `URISchemes` is now the same as we have. Therefore, we should adapt this rule, or remove it entirely if we want to stick to upstream (that being said, 100 columns is a good default if you ask me...).
- Remove everything else: each team is then free to add whatever rules they want, but this way we reduce the maintenance burden.

**Comment by cornelius** [comment]
Just my 2 cents: I think the changes @mssola proposes are good.

**Comment by Ana06** [comment]
@mssola 

> It really doesn't. From time to time we have to update this to catch up with upstream (see #17), which is troublesome.

That also means that everybody using this SUSE style have to wait for updates here to update Rubocop... 😭 

> Theoretically anyone from SUSE that cares about this. That being said, in practice not much time has been spent on this style guide, since everyone has more pressing concerns (see PRs that have stalled for a long time now...). So, it's a problem...

Why do we want to discuss inside SUSE about Ruby Style instead of doing it with the Ruby community in Rubocop? Rubocop is community maintain and it is in my opinion an example of how people work on open source. We 💚 open source at SUSE, so I would aim that we collaborate with Rubocop instead of having our own discussions.


> Keep stuff that we agree and the community guide has no strong opinions. For example: string literals. The community guide says that both options are valid, and when we discussed about it at SUSE, the preferred options seemed double quotes (which is something I didn't agree much at first, but now I got used to it 😅).

This has also been discussed in Rubocop (see https://github.com/bbatsov/rubocop/issues/5306), and they are actually trying to measure what people prefer. I think it would be nice to let them know what you think so that the change can be made.

**Comment by aspiers** [comment]
I think I agree with everything said here. Let's move as much of our code styling upstream as possible, since in general there's nothing magical about SUSE Ruby code which would require different style to the rest of the world, and then we can benefit from all the work in the upstream community rather than having the burden of maintaining yet another SUSE silo.

**Comment by mssola** [comment]
@Ana06 @aspiers @cornelius agreed with everything said here. Meanwhile in Portus we already moved to upstream rubocop (with some SUSE thingies like double-quotes). So, whenever I have some time I will:

1. Submit a PR to this repository with a more minimalist style (what I've said in [this comment](https://github.com/SUSE/style-guides/issues/18#issuecomment-348436540)).
2. Go to rubocop and give some feedback as @Ana06 suggested.

**Comment by Ana06** [comment]
@mssola 

> Meanwhile in Portus we already moved to upstream rubocop

Any of the projects I work at SUSE is following this... 😂  

> Submit a PR to this repository with a more minimalist style (what I've said in this comment).

what about get rid of this repository? 😉 

> Go to rubocop and give some feedback as @Ana06 suggested.

That would be great! ❤️

**Comment by mssola** [comment]
> what about get rid of this repository?

Well, we'd still need a reference configuration for stuff like double quotes.

**Comment by Ana06** [comment]
> Well, we'd still need a reference configuration for stuff like double quotes.

The one in upstream, we would have to discuss here if we want doubles or single quotes. Instead we can do it in upstream. And if a concrete team at SUSE can not live with that they can modify it in their projects. 😉

**Comment by bgeuken** [comment]
> aspiers commented 17 hours ago
> 
> I think I agree with everything said here. Let's move as much of our code styling upstream as possible, since in general there's nothing magical about SUSE Ruby code which would require different style to the rest of the world

It's also not feasible, since we work and contribute to many projects that are driven or influenced by people outside of SUSE:-)

I like @mssola's proposal

**Comment by mssola** [comment]
> we would have to discuss here if we want doubles or single quotes

We already discussed this some years ago (where there were people from all teams using ruby), and we agreed on some stuff (e.g. double quotes). That's why I was proposing a "softer" or a step-by-step way of deleting this repository (a.k.a move the discussion upstream but meanwhile we keep the stuff we agreed).

Of course, for me as a maintainer, it would be easier to simply move this conversation upstream and wipe out the repository right now. Would people agree to this ?

**Comment by Ana06** [comment]
@mssola 

> Of course, for me as a maintainer, it would be easier to simply move this conversation upstream and wipe out the repository right now. Would people agree to this ?

🎉 😍 💚

**Comment by aspiers** [comment]
@mssola Not everyone is following this issue, so I'd suggest first mailing ruby-devel@suse.de to give people a chance to object, as I guess there's a small chance that deleting the repo might break things somewhere. But once people have had fair notice, I'm in favour of deleting.

**Comment by suntorytimed** [comment]
So what's the state of this? I was redirected to this project by the rubocop configurations in our SAP yast modules. Would it be best to just use the Portus configuration or is there another, better configuration that should be used? If so please update the README and redirect to the proper configuration.

**Comment by Ana06** [comment]
@suntorytimed 

> So, what's the state of this? I was redirected to this project by the rubocop configurations in our SAP yast modules. Would it be best to just use the Portus configuration or is there another, better configuration that should be used? If so please update the README and redirect to the proper configuration.

In my opinion this repository should be deleted. It seems by the discussion in this issue that people agree, the only thing missed was writing to ruby-devel@suse.de. @aspiers have you done it already or should I do it? I am not subscribe to it, what is the use of ruby-devel? :thinking:

**Comment by Ana06** [comment]
@aspiers 

> I didn't - please go ahead and mail ruby-devel@suse.de :-)

Done :+1: 
> Since you are working on Ruby, please subscribe to this list. It's low traffic, and occasionally there is important information on there. The mailman page just says it's for "Discussion Ruby and Ruby On Rails" which is very vague, but IIUC it's basically for anyone inside SUSE who works with Ruby. Take a look at the archives to get a better feel for what is discussed.

Done :+1:

**Comment by aspiers** [comment]
Agreed!  Unfortunately I don't have permissions to do that myself.
